File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/ui/nativerender/NativeTreeObserverRegistry.kt
package com.nativephp.mobile.ui.nativerender

import java.util.concurrent.atomic.AtomicInteger

/** Opt-in observation seam for the decoded native tree. */
object NativeTreeObserverRegistry {
    data class Subscription internal constructor(internal val id: Int)

    private val sequence = AtomicInteger(0)
    private val lock = Any()
    private val observers = linkedMapOf<Int, (NativeUITree) -> Unit>()
    @Volatile private var latestTree: NativeUITree? = null
    @Volatile private var hasObservers = false

    fun observe(observer: (NativeUITree) -> Unit): Subscription {
        val id = sequence.incrementAndGet()
        val replay = synchronized(lock) {
            observers[id] = observer
            hasObservers = true
            latestTree
        }
        replay?.let(observer)
        return Subscription(id)
    }

    fun unsubscribe(subscription: Subscription) {
        synchronized(lock) {
            observers.remove(subscription.id)
            hasObservers = observers.isNotEmpty()
        }
    }

    internal fun publish(tree: NativeUITree) {
        latestTree = tree
        if (!hasObservers) return
        val current = synchronized(lock) { observers.values.toList() }
        current.forEach { observer -> runCatching { observer(tree) } }
    }
}

File: resources/xcode/NativePHP/NativeRender/NativeTreeObserverRegistry.swift
import Foundation

/// Opt-in observation seam for the decoded native tree.
final class NativeTreeObserverRegistry {
    static let shared = NativeTreeObserverRegistry()

    struct Subscription: Hashable { fileprivate let id: Int }

    private let lock = NSLock()
    private var sequence = 0
    private var observers: [Int: (NativeUITree) -> Void] = [:]
    private var latestTree: NativeUITree?

    private init() {}

    func observe(_ observer: @escaping (NativeUITree) -> Void) -> Subscription {
        lock.lock()
        sequence &+= 1
        let subscription = Subscription(id: sequence)
        observers[sequence] = observer
        let replay = latestTree
        lock.unlock()
        if let replay { observer(replay) }
        return subscription
    }

    func unsubscribe(_ subscription: Subscription) {
        lock.lock(); defer { lock.unlock() }
        observers.removeValue(forKey: subscription.id)
    }

    func publish(_ tree: NativeUITree) {
        lock.lock()
        latestTree = tree
        guard !observers.isEmpty else { lock.unlock(); return }
        let current = Array(observers.values)
        lock.unlock()
        for observer in current { observer(tree) }
    }
}
